### PR TITLE
[GSoC] Formula Editor Lists

### PR DESCRIPTION
**Pull request by obusher**
* Reopening of #1100 Pullrequest from @bernhardsp.
* Rebased to current master. 
* [Testrun (green)](https://jenkins.catrob.at/view/All-Categories/view/Catroid-multi-job/job/Catroid-Multi-Job-Custom-Branch-RELOADED/696/) :smile:

### CODE REVIEW

**Comment by SimonStefan** [comment]
Could you please post a gist of the code.xml that is created when a project with lists is saved?

**Comment by bernhardsp** [comment]
https://gist.github.com/bernhardsp/a2d53b6067ab6a74a882

The project contains a local and a global list.

**Comment by avihric** [comment]
Could you rebase to master and squash your commits

**Comment by obusher** [comment]
The PR-test fails now after rebase onto `master`.
@bernhardsp already squashed the commits already to a sensible amount.
( ~20 commits for this **big** PR schould be ok. :) )

**Comment by obusher** [comment]
Current [testrun] (https://jenkins.catrob.at/view/All-Categories/view/Catroid-multi-job/job/Catroid-Multi-Job-Custom-Branch-RELOADED/779/) is yellow with [1 fail](https://github.com/Catrobat/Catroid/blob/master/catroidTest/src/org/catrobat/catroid/uitest/ui/activity/UserBrickScriptActivityTest.java#L74) .  The failing test passes on my device with no problems. :)

@SimonStefan did you check the code.xml?

**Comment by SimonStefan** [comment]
yes, i made a comment there

**Comment by obusher** [comment]
I do not get any notification if someone is commenting the [.xml-gist](https://gist.github.com/bernhardsp/a2d53b6067ab6a74a882), so i will continue here:

* `<objectListOfList\>` :         list of local userLists 
* `<objectVariableList\>` :        list of local useVariables
* `<programListOfLists\>` :        list of global userLists 

`<objectVariableList\>` is already introduced. 
The other 2 elements we need to distinguish between object (sprite) and program userlists (~scope).

**Comment by robertpainsi** [comment]
Since there is a backspace button on the keyboard layout now, a new problem occurred.
Delete all the content in the formula. Now the formula is empty. Press the backspace button on the keyboard again, a NullPointerException is thrown.

```
03-07 23:05:41.861  15387-15387/org.catrobat.catroid E/MessageQueue-JNI﹕ java.lang.NullPointerException: Attempt to invoke virtual method 'int org.catrobat.catroid.formulaeditor.InternFormula$CursorTokenPosition.ordinal()' on a null object reference
            at org.catrobat.catroid.formulaeditor.InternFormula.handleDeletion(InternFormula.java:292)
            at org.catrobat.catroid.formulaeditor.InternFormula.handleKeyInput(InternFormula.java:114)
            at org.catrobat.catroid.formulaeditor.FormulaEditorEditText.handleKeyEvent(FormulaEditorEditText.java:234)
            at org.catrobat.catroid.ui.fragment.FormulaEditorFragment$1.onTouch(FormulaEditorFragment.java:300)
            at android.view.View.dispatchTouchEvent(View.java:8382)
            ...
```

Disable/enable the backspace button like in the edit field.

**Comment by obusher** [comment]
NullPointerException fixed.
New Testrun is [green](https://jenkins.catrob.at/job/Catroid-ParalellEmulatorUITest-CustomBranch/884/). :)

 3..........2........1..........MERGE!

File: catroid/src/org/catrobat/catroid/content/bricks/AddItemToUserListBrick.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+	public AddItemToUserListBrick(Formula userListFormula, UserList userList) {
+		initializeBrickFields(userListFormula);
+		this.userList = userList;
+	}
+
+	public AddItemToUserListBrick( double value) {
+		initializeBrickFields(new Formula(value));
+	}
+
+	@Override
```
Comment (SimonStefan) [inline catroid/src/org/catrobat/catroid/content/bricks/AddItemToUserListBrick.java]: That's standard behavior. Don't override it.

File: catroid/src/org/catrobat/catroid/content/bricks/DeleteItemOfUserListBrick.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	@Override
+	public Brick clone() {
+		DeleteItemOfUserListBrick clonedBrick = new DeleteItemOfUserListBrick(getFormulaWithBrickField(BrickField.LIST_DELETE_ITEM).clone(), userList);
+		return clonedBrick;
+	}
+
+	@Override
+	public void onClick(View view) {
```
Comment (SimonStefan) [inline catroid/src/org/catrobat/catroid/content/bricks/DeleteItemOfUserListBrick.java]: standard behavior?

**Comment by SimonStefan** [comment]
1.) We must get rid of the reference parts in the xml! I can't see a reason why the brick should hold a reference to the concrete list object. We already have the same problem with the variables. e.g. you could work with IDs or with name and type.

2.) The brick should have nothing to do with the NewDataDialog. (probably for later refactoring)

3.) The copyBrickForSprite() in every Brick is a huge object-oriented fault and you shouldn't depend on it. If you fix 1., clone() can be used everywhere you need this method. And you probably wont need to override it in every class.

4.) The LIST_DELETE_ITEM is a INDEX ?


[1] https://gist.github.com/bernhardsp/a2d53b6067ab6a74a882#file-project-with-global-and-local-list-L63-L71

**Comment by robertpainsi** [comment]
@obusher thx for fixing it that fast!

**Comment by linsm** [comment]
ok to test

**Comment by X4mp** [comment]
test this please

**Comment by obusher** [comment]
@robertpainsi :  Your welcome! :)

**Pull request closed without merging**